CoViewingExample/AudioDevices/ExampleAVPlayerProcessingTap.m

```diff
@@ -0,0 +1,373 @@
+        context->sourceCache = NULL;
+        context->sourceCacheFrames = 0;
+    }
+
+    if (context->renderFormatConverter != NULL) {
+        AudioConverterDispose(context->renderFormatConverter);
+        context->renderFormatConverter = NULL;
+    }
+
+    if (context->captureFormatConverter != NULL) {
+        AudioConverterDispose(context->captureFormatConverter);
+        context->captureFormatConverter = NULL;
+        context->captureFormatConvertIsPrimed = NO;
+    }
+}
+
+void AVPlayerProcessingTapProcess(MTAudioProcessingTapRef tap,
+                                  CMItemCount numberFrames,
+                                  MTAudioProcessingTapFlags flags,
+                                  AudioBufferList *bufferListInOut,
+                                  CMItemCount *numberFramesOut,
+                                  MTAudioProcessingTapFlags *flagsOut) {
+    ExampleAVPlayerAudioTapContext *context = (ExampleAVPlayerAudioTapContext *)MTAudioProcessingTapGetStorage(tap);
+    CMTimeRange sourceRange;
+    OSStatus status = MTAudioProcessingTapGetSourceAudio(tap,
+                                                         numberFrames,
+                                                         bufferListInOut,
+                                                         flagsOut,
+                                                         &sourceRange,
+                                                         numberFramesOut);
+
+    if (status != kCVReturnSuccess) {
+        // TODO
+        return;
+    }
+
+    UInt32 framesToCopy = (UInt32)*numberFramesOut;
+
+    // Produce renderer buffers. These are interleaved, signed integer frames in the source's sample rate.
+    TPCircularBuffer *renderingBuffer = context->renderingBuffer;
+    AVPlayerAudioTapProduceConvertedFrames(renderingBuffer, context->renderFormatConverter, bufferListInOut, framesToCopy, 2);
+
+    // Produce capturer buffers. We will perform a sample rate conversion if needed.
+    TPCircularBuffer *capturingBuffer = context->capturingBuffer;
+    if (context->capturingSampleRateConversion) {
+        AVPlayerAudioTapProduceFilledFrames(capturingBuffer, context->captureFormatConverter, context->captureFormatConvertIsPrimed, bufferListInOut, context->sourceCache, &context->sourceCacheFrames, framesToCopy, kPreferredNumberOfChannels);
+        context->captureFormatConvertIsPrimed = YES;
+    } else {
+        AVPlayerAudioTapProduceConvertedFrames(capturingBuffer, context->captureFormatConverter, bufferListInOut, framesToCopy, kPreferredNumberOfChannels);
+    }
+
+    // Flush converters on a discontinuity. This is especially important for priming a sample rate converter.
+    if (*flagsOut & kMTAudioProcessingTapFlag_EndOfStream) {
+        AudioConverterReset(context->renderFormatConverter);
+        AudioConverterReset(context->captureFormatConverter);
+        context->captureFormatConvertIsPrimed = NO;
+    }
+}
```

Comment:
This file has format conversion code as well. We will be using `ExampleAVPlayerProcessingTap.m` in AVAudioEngine audio device as well where the format conversion will not be required. Can we move the audio format conversion code outside this file? another option is we can have processing tap per audio device, in that case we will need to add a layer in folder structure inside `AudioDevices`

Reply:
Also I noticed, audio capturing side stops working when app goes into the background. It resumes coming into the foreground.